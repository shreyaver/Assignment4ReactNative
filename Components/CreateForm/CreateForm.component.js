/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow
 * @lint-ignore-every XPLATJSCOPYRIGHT1
 */

import React, { Component } from 'react';
import {
  SafeAreaView, Text, TouchableOpacity, TextInput, ScrollView, View, Platform,
} from 'react-native';
import PropTypes from 'prop-types';
import styles from './CreateForm.component.stylesheet';
import Icon from 'react-native-vector-icons/Entypo';
import constants from '../../config';
import Axios from 'axios';

export default class CreateForm extends Component {
  static navigationOptions = {
    title: 'NewForm',
  };

  state = { 
    name: '',
    fields: [],
    status: '',
  };
  
  handleSubmit = async () => {
    const { name, fields } = this.state;
    if (name === '' || /^\s+$/.test(name)) {
      this.setState({ status: constants.NAME_BLANK_ERROR_MESSAGE });
      return;
    }
    if (!/[a-z]|[0-9]+/i.test(name)) {
      this.setState({ status: constants.NAME_NO_NUMBER_OR_ALPHABET_ERROR_MESSAGE });
      return;
    }
    if (fields.length === 0) {
      this.setState({ status: constants.FIELDS_EMPTY_ARRAY_ERROR_MESSAGE });
      return;
    }
    if (fields.some(field => field === '' || /^\s+$/.test(field))) {
      this.setState({ status: constants.FIELDS_BLANK_ERROR_MESSAGE });
      return;
    }
    if (fields.some(field => !/[a-z]|[0-9]+/i.test(field))) {
      this.setState({ status: constants.FIELDS_NO_NUMBER_OR_ALPHABET_ERROR_MESSAGE });
      return;
    }
    console.log('hello');
    const insertResult = await Axios.post(Platform.OS === 'ios' ? constants.POST_FORM_URL_IOS : constants.POST_FORM_URL_ANDROID, { name, fields });
    if (insertResult.data.message !== 'Created form') {
      this.setState({ status: 'Unable to create form' });
      return;
    } else {
      this.props.navigation.navigate('Home');
    }
  }
  render() {
    return (
      <SafeAreaView style={styles.safeAreaView}>
        <View style={styles.header}>
          <Text style={styles.heading}> Create Form </Text>   
          <Icon
            name="dots-three-vertical"
            color="white"
            size={40}
            style={styles.dotIcon}
          />
        </View>       
        <ScrollView contentContainerStyle={styles.screenContent}>
          <View style={styles.containerView}>
            <TextInput
              style={styles.formNameInput}
              multiline
              maxLength={250}
              selectionColor="black"
              placeholder="Form Name"
              testID="formName"
              onChangeText={(newName) => { this.setState({ name: newName }); }}
            />
            <Text>
              {this.state.status}
            </Text>
            <TouchableOpacity style={styles.addFieldButton} testID="addFieldButton" onPress={() => {
              this.setState((state) => ({ fields: [...state.fields, ''] }));
              }}
            >
              <Text style={styles.buttonText}>
                ADD FIELD
              </Text>
            </TouchableOpacity>
            <View style={styles.fieldsView} testID="fieldsView">
              {this.state.fields.map((value, index) => (
                <TextInput
                  style={styles.input}
                  multiline
                  maxLength={250}
                  selectionColor="black"
                  key={index}
                  value={this.state.fields[index]}
                  onChangeText={(newField) => { 
                    this.setState((state) => ({ fields: [...state.fields.slice(0, index), newField, ...state.fields.slice(index + 1)] })); 
                  }}
                />))
              }
            </View>
          </View>
          <TouchableOpacity style={styles.saveButton} testID="submitButton" onPress={this.handleSubmit}>
            <Text style={styles.buttonText}>
              SAVE
            </Text>
          </TouchableOpacity>
        </ScrollView>
      </SafeAreaView>
    );
  }
}
CreateForm.propTypes = {
  navigation: PropTypes.shape({ navigate: PropTypes.func.isRequired }).isRequired,
};