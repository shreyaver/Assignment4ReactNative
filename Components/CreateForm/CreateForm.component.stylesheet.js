import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  safeAreaView: {
    flex: 1,
    backgroundColor: 'transparent',
  },
  header: {
    height: 90,
    width: '100%',
    backgroundColor: '#3F51B5',
    flexDirection: 'row',
    alignItems: 'center',
  },
  heading: {
    color: 'white',
    fontSize: 22,
    marginLeft: 20,
  },
  dotIcon: {
    marginLeft: 'auto',
  },
  screenContent: {
    width: '100%',
    flexGrow: 1,
    paddingTop: 30,
    backgroundColor: 'white',
    alignItems: 'center',
  },
  containerView: {
    width: '85%',
  },
  input: {
    borderColor: 'grey',
    borderBottomWidth: 1,
    fontSize: 20,
    marginBottom: 20,
  },
  formNameInput: {
    borderColor: 'grey',
    borderBottomWidth: 1,
    fontSize: 20,
    marginBottom: 10,
  },
  addFieldButton: {
    marginTop: 10,
    width: 120,
    alignSelf: 'flex-end',
    alignItems: 'center',
    paddingVertical: 10,
    backgroundColor: '#3F51B5',
  },
  buttonText: {
    fontSize: 18,
    color: 'white',
  },
  fieldsView: {
    paddingVertical: 20,
  },
  saveButton: {
    width: '100%',
    height: 70,
    position: 'absolute',
    bottom: 0,
    backgroundColor: '#3F51B5',
    justifyContent: 'center',
    alignItems: 'center',
  },
});
