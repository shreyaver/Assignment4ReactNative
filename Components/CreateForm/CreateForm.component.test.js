import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Axios from 'axios';
import CreateForm from './CreateForm.component';
import constants from '../../config';
import { TextInput } from 'react-native';


describe('CreateForm', () => {
  let navigation;
  beforeAll(() => {
    navigation = { navigate: jest.fn() };
    mockGetForms = jest.spyOn(Axios, 'post');
    mockGetForms.mockImplementation(() => Promise.resolve({ data: 'Created form' }));
  });
  afterAll(() => {
    jest.restoreAllMocks();
  });
  it('renders without crashing', () => {
    const createForm = renderer.create(<CreateForm navigation={navigation} />).toJSON();
    expect(createForm).toMatchSnapshot();
  });
  it('changes "name" state to text when text is entered in "name" field', () => {
    const wrapper = shallow(<CreateForm navigation={navigation} />);
    expect(wrapper.instance().state.name).toEqual('');
    wrapper.find({ testID: 'formName' }).simulate('changeText', 'hello');
    expect(wrapper.instance().state.name).toEqual('hello');
  });
  it('adds new form field on clicking "add field" button', () => {
    const wrapper = shallow(<CreateForm navigation={navigation} />);
    expect(wrapper.instance().state.fields).toEqual([]);
    wrapper.find({ testID: 'addFieldButton' }).simulate('press');
    expect(wrapper.instance().state.fields).toEqual(['']);
  });
  it('changes "fields" state when text is entered in "form field name"', () => {
    const wrapper = shallow(<CreateForm navigation={navigation} />);
    wrapper.find({ testID: 'addFieldButton' }).simulate('press');
    wrapper.find({ testID: 'fieldsView' }).find(TextInput).simulate('changeText', 'hello');
    expect(wrapper.instance().state.fields).toEqual(['hello']);
  });
  it('should give error on inserting string with only blank spaces in "name"', () => {
    const formObj = {
      name: ' ',
      fields: ['First Name', 'Last Name', 'Address', 'Country'],
    };
    const wrapper = shallow(<CreateForm navigation={navigation} />);
    wrapper.instance().setState(formObj);
    wrapper.find({ testID: 'submitButton' }).simulate('press');
    expect(wrapper.instance().state.status).toEqual(constants.NAME_BLANK_ERROR_MESSAGE);
  });
  it('should give error on inserting string without any alphabet or number in "name"', () => {
    const formObj = {
      name: '.',
      fields: ['First Name', 'Last Name', 'Address', 'Country'],
    };
    const wrapper = shallow(<CreateForm navigation={navigation} />);
    wrapper.instance().setState(formObj);
    wrapper.find({ testID: 'submitButton' }).simulate('press');
    expect(wrapper.instance().state.status).toEqual(constants.NAME_NO_NUMBER_OR_ALPHABET_ERROR_MESSAGE);
  });
  it('should give error on inserting empty array in "fields"', () => {
    const formObj = {
      name: 'Personal Details',
      fields: [],
    };
    const wrapper = shallow(<CreateForm navigation={navigation} />);
    wrapper.instance().setState(formObj);
    wrapper.find({ testID: 'submitButton' }).simulate('press');
    expect(wrapper.instance().state.status).toEqual(constants.FIELDS_EMPTY_ARRAY_ERROR_MESSAGE);
  });
  it('should give error on inserting string with only blank spaces as a field', () => {
    const formObj = {
      name: 'Personal Details',
      fields: [' ', 'Last Name', 'Address', 'Country'],
    };
    const wrapper = shallow(<CreateForm navigation={navigation} />);
    wrapper.instance().setState(formObj);
    wrapper.find({ testID: 'submitButton' }).simulate('press');
    expect(wrapper.instance().state.status).toEqual(constants.FIELDS_BLANK_ERROR_MESSAGE);
  });
  it('should give error on inserting string without any alphabet or number as a field', () => {
    const formObj = {
      name: 'Personal Details',
      fields: ['.', 'Last Name', 'Address', 'Country'],
    };
    const wrapper = shallow(<CreateForm navigation={navigation} />);
    wrapper.instance().setState(formObj);
    wrapper.find({ testID: 'submitButton' }).simulate('press');
    expect(wrapper.instance().state.status).toEqual(constants.FIELDS_NO_NUMBER_OR_ALPHABET_ERROR_MESSAGE);
  });
});
