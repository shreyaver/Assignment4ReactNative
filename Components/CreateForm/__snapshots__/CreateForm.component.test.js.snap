// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateForm renders without crashing 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#3F51B5",
        "flexDirection": "row",
        "height": 90,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 22,
          "marginLeft": 20,
        }
      }
    >
       Create Form 
    </Text>
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 40,
          },
          Object {
            "marginLeft": "auto",
          },
          Object {
            "fontFamily": "Entypo",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      ï†‚
    </Text>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "flexGrow": 1,
        "paddingTop": 30,
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "width": "85%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          maxLength={250}
          multiline={true}
          onChangeText={[Function]}
          placeholder="Form Name"
          selectionColor="black"
          style={
            Object {
              "borderBottomWidth": 1,
              "borderColor": "grey",
              "fontSize": 20,
              "marginBottom": 10,
            }
          }
          testID="formName"
          underlineColorAndroid="transparent"
        />
        <Text>
          
        </Text>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "flex-end",
              "backgroundColor": "#3F51B5",
              "marginTop": 10,
              "opacity": 1,
              "paddingVertical": 10,
              "width": 120,
            }
          }
          testID="addFieldButton"
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            ADD FIELD
          </Text>
        </View>
        <View
          style={
            Object {
              "paddingVertical": 20,
            }
          }
          testID="fieldsView"
        />
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#3F51B5",
            "bottom": 0,
            "height": 70,
            "justifyContent": "center",
            "opacity": 1,
            "position": "absolute",
            "width": "100%",
          }
        }
        testID="submitButton"
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
            }
          }
        >
          SAVE
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
